src/DictList.elm

```diff
@@ -938,6 +965,174 @@ fromDict dict =
 
 
 
+-- LIST EXTRA
+
+
+{-| Extract the last element of a list.
+    last (fromList [(1, 1), (2, 2), (3, 3)]) == Just (3, 3)
+    last (fromList []) == Nothing
+-}
+last : DictList comparable v -> Maybe ( comparable, v )
+last xs =
+  toList xs
+  |> List.Extra.last
+
+
+{-| Return all initial segments of a list, from shortest to longest, empty list first, the list itself last.
+
+    inits (fromList [(1, 1),(2,),(3, 3)]) == [fromList [], fromList [(1, 1)], fromList [(1, 1), (2, 2)], fromList [(1, 1), (2, 2), (3, 3)]]
+-}
+inits : DictList comparable v -> List (DictList comparable v)
+-- @FIXME
+inits list =
+  toList list
+    |> List.Extra.inits
+    |> List.map fromList
+
+{-| Returns a list of repeated applications of `f`.
+
+If `f` returns `Nothing` the iteration will stop. If it returns `Just y` then `y` will be added to the list and the iteration will continue with `f y`.
+    nextYear : Int -> Maybe Int
+    nextYear year =
+      if year >= 2030 then
+        Nothing
+      else
+        Just (year + 1)
+    -- Will evaluate to [2010, 2011, ..., 2030]
+    iterate nextYear 2010
+-}
+iterate : ((comparable, v) -> Maybe (comparable, v)) -> (comparable, v) -> DictList comparable v
+iterate f x =
+  List.Extra.iterate f x
+  |> fromList
+
+{-| Decompose a list into its head and tail. If the list is empty, return `Nothing`. Otherwise, return `Just (x, xs)`, where `x` is head and `xs` is tail.
+
+    uncons (fromList [(1, 1),(2, 2),(3, 3)] == Just ((1, 1), fromList [(2, 2), (3, 3)])
+    uncons empty = Nothing
+-}
+uncons : DictList comparable v -> Maybe ( (comparable, v), DictList comparable v)
+uncons xs =
+  toList xs
+    |> List.Extra.uncons
+    |> Maybe.map (\(a, la) -> (a, fromList la))
+
+{-| Find the first maximum element in a list using a comparable transformation
+-}
+maximumBy : (comparable2 -> a -> comparable1) -> DictList comparable2 a -> Maybe (comparable2, a)
+maximumBy f ls =
+  toList ls
+    |> List.Extra.maximumBy (uncurry f)
```

Comment:
This one looks fine to me -- I don't think there's an optimization possible using the internals, since we'll have to get all the values one way or another. I suppose we might be able to avoid constructing the intermediate list by using a `foldl` (and then using something roughly like the implementation of `List.Extra.maximumBy`).